### PR TITLE
HIVE-24694: Early connection close to release server resources during creating

### PR DESCRIPTION
**Pull request by dengzhhu653**

### CODE REVIEW

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 1]
```diff
@@ -357,6 +357,15 @@ public HiveConnection(String uri, Properties info) throws SQLException {
           }
 
           if (numRetries >= maxRetries) {
+            try {
```
Comment (kgyrtkirk) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:360]: I was looking around and made the following notices:
* this check only occurs right before we throw the exception out
* what will happen with opened connection in between retries?
* it seems to me that `openTransport` does a check on the transport that is open or not - but it creates a new transport anyway; so I assume the driver will open a fresh connection for every try

I'm thinking the following:
* what do you think about closing the connection after a failed attempt?

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 2]
```diff
@@ -357,6 +357,15 @@ public HiveConnection(String uri, Properties info) throws SQLException {
...
+              // Cleaning up the server resources early
+              if (!isClosed) {
+                // close the session
+                close();
+              } else if (transport != null) {
+                transport.close();
+              }
```
Comment (kgyrtkirk) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:367]: I think you could also extend the close to be sure to close the transport if that's the only one open

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java (hunk 1 above)
Comment (dengzhhu653) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:360]: yes, you're right. Thanks very much for the advice and comments!

File: jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java [hunk 3]
```diff
@@ -1104,18 +1102,18 @@ public void clearWarnings() throws SQLException {
...
+          client.CloseSession(closeReq);
+        }
       } catch (TException e) {
         throw new SQLException("Error while cleaning up the server resources", e);
       } finally {
         isClosed = true;
-        if (transport != null) {
+        if (transport != null && transport.isOpen()) {
           transport.close();
         }
```
Comment (kgyrtkirk) [inline jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:1116]: can we release the references as well?
```
transport=null
client=null
```

Comment (dengzhhu653) [reply jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java:1126]: done and add some pre-checks on the connection state to avoid potential NPE

**Review by kgyrtkirk — APPROVED** [review]

**Pull request merged**